### PR TITLE
chore: Vendor subset of dependencies to avoid namespace collisions

### PR DESCRIPTION
**Pull request by Sambigeara**
Use third-party distributions for missing proto-generated dependencies where available, thus reducing the chance of namespace collisions downstream.

Closes https://github.com/cerbos/cerbos-sdk-python/issues/44

### CODE REVIEW

**Comment by Sambigeara** [comment]
The final hurdle here is that it requires the users to specify and allow an additional repository in their configuration, e.g:

```toml
[[tool.pdm.source]]
name = "buf"
url = "https://buf.build/gen/python"
```

If they don't do this, the installation step will fail. The alternative method here is hilariously to vendor the dependencies ourselves (which is the very thing I'm trying to avoid). A little bit chicken<->egg.

Rather than requiring users to explicitly add the repo config to their projects, I wonder if we could instead mirror the wheels built by Buf and host them in our Pypi repository 🤔.

Alternatively, I finally give in and publish a "legacy" `http` package separate from the main branch.

@charithe - wondering if you have any thoughts on this?

**Comment by charithe** [comment]
I am a bit reluctant to have the BSR as a critical component in the user-facing distribution path because it's unclear whether it'll be free forever. Furthermore, the fact that users have to configure a new repository for their Python package manager adds a hurdle because some environments have strict rules about which third-party repositories are allowed.

If we can pull the wheels from BSR and distribute them as part of our Python package, that would be ideal. They have to be downloaded automatically as part of our package (without any additional configuration on the user side) though.

**Comment by Sambigeara** [comment]
I managed to get it all running, vendoring the wheels directly and referencing as local dependencies, got to the point of pushing to the test PyPI index, and... 🥁

```sh
INFO     Response from https://test.pypi.org/legacy/:
         400 Invalid value for requires_dist. Error: Can't have direct dependency:
         'cerbos-cerbos-api-grpc-python@
         file:///Users/samuellock/cerbos/cerbos-sdk-python/libs/cerbos_cerbos_api_grpc_python-1.60.0.1.202
         40110153040%2B75004b0044b2-py3-none-any.whl'
```

😞. A bit of digging found a similar issue, and [this particular comment](https://github.com/pypi/warehouse/issues/7136#issuecomment-603529974):

> PEP 440 says:
> 
> > Public index servers SHOULD NOT allow the use of direct references in uploaded distributions.
> 
> PEP 508 says:
> 
> > Firstly, PEP-440 direct references haven't actually been deployed in the wild, but they were designed to be compatibly added, and there are no known obstacles to adding them to pip or other tools that consume the existing dependency metadata in distributions - particularly since they won't be permitted to be present in PyPI uploaded distributions anyway.
> 
> As such, I'm going to close this issue as we have no plans to deviate from these PEPs, but thanks for the report.

_Sigh_

**Comment by charithe** [comment]
That's frustrating 😞 

Taking a step back, I am confused about the root cause of #44 because I haven't done any advanced Python packaging. Am I right in saying that:
- the protobuf code we generate has references to Google protobuf generated code
- those references are not namespaced (because Python protobuf generator doesn't support that)
- including the google dependencies in our package causes problems for users who already have a dependency on them through another route

If the above is correct:
- how does the Buf generated code work around it? can we replicate it locally?
- if we produce a package variant that doesn't require google protos, will that address #44?

**Comment by Sambigeara** [comment]
Yeah your assumptions seem correct.

I think the crux of the issue was that `googleapis-common-protos` had the same namespace as the `google` code built via `buf generate` due to our own proto definition's dependency on it. This, coupled with the fact that the import paths generated in the code were absolute and strictly matched the proto module structure.

I'm not entirely sure why, but importing the Buf generated SDKs from their index resolved #44 , I'm initially thought due to smarter dependency resolution, but on further thought it could be due to [namespace package magic](https://packaging.python.org/en/latest/guides/packaging-namespace-packages/) or something else I'm unfamiliar with (namespace packages seem to allow for duplicate namespaces, providing there are no collisions in submodules..?).

> if we produce a package variant that doesn't require google protos, will that address https://github.com/cerbos/cerbos-sdk-python/issues/44?

I think so. Our other generated namespaces are less likely to collide than `google` 😆.

**Comment by charithe** [comment]
Thanks! I think I (kind of) get it now. What if we use the [buf option to restrict the set of generated types](https://buf.build/docs/generate/overview#limit-to-specific-types) to avoid generating code for `google` and `protoc_gen_openapiv2` and instead add dependencies on `googleapis-common-protos` and `protoc-gen-openapiv2-protoc3-19`?

**Comment by charithe** [comment]
Could we create a namespace package for `google.api.expr`? There is some precedent for grpc-gateway openapiv2 protos (https://github.com/unionai-oss/protoc-gen-openapiv2).

**Comment by Sambigeara** [comment]
Yeah I think that works. If I add the two additional dependencies (`googleapis-common-protos` and `protoc-gen-openapiv2`), delete all generated dependency proto code _apart_ from `google/api/expr/`, then the collision issue seems to go away. Perhaps this is the way to go for now, given there don't seem to be any other viable options.

`buf generate` type filters don't seem to work on imported modules:

```sh
Failure: filtering by type "google.api.expr.v1alpha1.CheckedExpr": type declared in imported module
```

So the way to approach this will be to generate all code into a temp directory and copy what we need into the top level 🤪. It's bonkers, the gymnastics (and ongoing maintenance costs) that have to be done to solve a seemingly simple issue!

**Comment by charithe** [comment]
> So the way to approach this will be to generate all code into a temp directory and copy what we need into the top level

I think we can make it work by  removing the `--include-imports` flag from https://github.com/cerbos/cerbos-sdk-python/blob/341f84297f79e3e8a9ad79966779f446a262fbba/proto/generate_protos.sh#L12 and then adding the following line to the script:

```
buf generate --path google/api/expr buf.build/googleapis/googleapis
```

**Comment by Sambigeara** [comment]
Yes! That works! Thanks so much for all the help. I went slightly off piste and opted for a two-step `buf export` -> `buf generate` step. It means that we filter out unused generated types. Hopefully should have a final commit incoming shortly.

**Review by charithe — COMMENTED** [review]
Pointed out a couple of small issues below. LGTM otherwise.

File: proto/generate_protos.sh
[the remark this answers is not in the document]
Comment (Sambigeara) [reply proto/generate_protos.sh:1]: TIL about `shellcheck` and also how you can run it in nvim via `efm-langserver` and now 🌈 🦄

[the remark this answers is not in the document]
Comment (Sambigeara) [reply proto/generate_protos.sh:1]: Also, I knew I was bad at bash but I didn't know I was _that_ bad

File: proto/generate_protos.sh [hunk 1]
```diff
@@ -1,14 +1,67 @@
...
+
+    echo ">> Import complete!"
+
+    echo ">> Tracking generated files..."
+
+    rm $CURRENT_PROTOS
+    echo "# DO NOT EDIT. This is generated automatically and used to track proto-generated code" >> $CURRENT_PROTOS
+    for file in $(find $TEMP_DEFS_DIR -type f); do
+        echo $(dirname ${file#$TEMP_DEFS_DIR/}) >> $CURRENT_PROTOS
+    done
```
[the remark this answers is not in the document]
Comment (charithe) [reply proto/generate_protos.sh]: Sorry, I got my wires crossed and read the last section as if it was taking the `--path` parameter from the list of files. Apparently I need new glasses 👓

**Review by charithe — APPROVED** [review]

File: buf/validate/expression_pb2.pyi [hunk 2]
```diff
@@ -0,0 +1,34 @@
+from google.protobuf.internal import containers as _containers
+from google.protobuf import descriptor as _descriptor
+from google.protobuf import message as _message
+from typing import ClassVar as _ClassVar, Iterable as _Iterable, Mapping as _Mapping, Optional as _Optional, Union as _Union
```
Comment (kojiromike) [inline buf/validate/expression_pb2.pyi:4]: Sorry for the OT, but I've never seen a stub file with all the imports aliased to have an underscore prefix. I would guess it has something to do with explicitly not re-exporting the name, but afaiu that should not be necessary, so I'm curious to learn if there's another purpose.

Comment (Sambigeara) [reply buf/validate/expression_pb2.pyi:4]: > I would guess it has something to do with explicitly not re-exporting the name

I would have guessed the same, but also couldn't be sure 🤔. I initially thought `buf generate` used `protoc` under the hood, but after a quick look, I'm not so sure. Regardless, I think the answer will be tucked away in the source of one of those somewhere. Sorry I couldn't be of more help!

**Pull request merged**